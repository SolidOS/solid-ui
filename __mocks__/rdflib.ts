export const Util =  {
  mediaTypeClass: sym
}

export function sym () {
  return {
    dir: sym,
    uri: 'uri',
    value: '',
    elements: [],
    doc: () => {
      return {
        equals: () => {},
        value: ''
      }
    },
    sameTerm: () => false
  }  
}  

export const uri = {
  join: () => {}
}
export function st() {
}

export function literal() {
}
export function graph() {
  return {
    any: () => {
      return sym()
    },
    each: () => {
      return []
    },
    match: () => {
      return []
    },
    query: () => {
      return []
    },
    fetcher: {
      load: () => {},
      nowOrWhenFetched: () => Promise.resolve()
    },
    findTypeURIs: () => {
      return []
    },
    findSuperClassesNT: () => {
      return []
    },
    bottomTypeURIs: () => {
      return []
    },
    statementsMatching: () => {
      return []
    },
    sym
  }
}
export function fetcher() {
  return {}
}
export class Fetcher {
  requested: any
  constructor() {
    this.requested = {}
  }
  load () {
  }
}
export function namedNode() {
  return {}
}
export function NamedNode() {
}
export function Namespace() {
  return () => {}
}
export class UpdateManager {
  editable() {}
  update() {}
}
