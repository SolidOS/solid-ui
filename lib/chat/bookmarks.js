"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findBookmarkDocument = findBookmarkDocument;
exports.renderBookmarksButton = renderBookmarksButton;
exports.toggleBookmark = toggleBookmark;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var debug = _interopRequireWildcard(require("../debug"));
var _iconBase = require("../iconBase");
var _index = require("../media/index");
var ns = _interopRequireWildcard(require("../ns"));
var pad = _interopRequireWildcard(require("../pad"));
var rdf = _interopRequireWildcard(require("rdflib"));
var style = _interopRequireWildcard(require("../style"));
var utils = _interopRequireWildcard(require("../utils"));
var widgets = _interopRequireWildcard(require("../widgets"));
var _solidLogic = require("solid-logic");
var _login = require("../login/login");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, "default": e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t4 in e) "default" !== _t4 && {}.hasOwnProperty.call(e, _t4) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t4)) && (i.get || i.set) ? o(f, _t4, i) : f[_t4] = e[_t4]); return f; })(e, t); }
/**
 * Functions related to chat and bookmarks
 * @packageDocumentation
 */

// pull in first avoid cross-refs

var UI = {
  icons: _iconBase.icons,
  ns: ns,
  media: _index.media,
  pad: pad,
  rdf: rdf,
  style: style,
  utils: utils,
  widgets: widgets
};
var $rdf = UI.rdf;
var BOOK = $rdf.Namespace('http://www.w3.org/2002/01/bookmark#');
var BOOKMARK_ICON = 'noun_45961.svg';
var label = utils.label;
var dom = window.document || null;

// @@@@ use the one in rdflib.js when it is avaiable and delete this
function updatePromise(del, ins) {
  return new Promise(function (resolve, reject) {
    _solidLogic.store.updater.update(del, ins, function (uri, ok, errorBody) {
      if (!ok) {
        reject(new Error(errorBody));
      } else {
        resolve();
      }
    }); // callback
  }); // promise
}

/*         Bookmarking
 */
/** Find a user's bookmarks
 */
function findBookmarkDocument(_x) {
  return _findBookmarkDocument.apply(this, arguments);
}
/** Add a bookmark
 */
function _findBookmarkDocument() {
  _findBookmarkDocument = (0, _asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function _callee(userContext) {
    var theClass, fileTail, isPublic, newBookmarkFile, _t;
    return _regenerator["default"].wrap(function (_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          theClass = BOOK('Bookmark');
          fileTail = 'bookmarks.ttl';
          isPublic = true;
          _context.next = 1;
          return (0, _login.findAppInstances)(userContext, theClass, isPublic);
        case 1:
          if (!(userContext.instances && userContext.instances.length > 0)) {
            _context.next = 2;
            break;
          }
          userContext.bookmarkDocument = userContext.instances[0];
          if (userContext.instances.length > 1) {
            debug.warn('More than one bookmark file! ' + userContext.instances); // @@ todo - deal with > 1
            // Note: should pick up community bookmarks as well
          }
          _context.next = 9;
          break;
        case 2:
          if (!userContext.publicProfile) {
            _context.next = 8;
            break;
          }
          // publicProfile or preferencesFile
          newBookmarkFile = $rdf.sym(userContext.publicProfile.dir().uri + fileTail);
          _context.prev = 3;
          debug.log('Creating new bookmark file ' + newBookmarkFile);
          _context.next = 4;
          return _solidLogic.store.fetcher.createIfNotExists(newBookmarkFile);
        case 4:
          _context.next = 6;
          break;
        case 5:
          _context.prev = 5;
          _t = _context["catch"](3);
          debug.warn("Can't make fresh bookmark file:" + _t);
          return _context.abrupt("return", userContext);
        case 6:
          _context.next = 7;
          return (0, _solidLogic.registerInTypeIndex)(newBookmarkFile, userContext.index, theClass);
        case 7:
          userContext.bookmarkDocument = newBookmarkFile;
          _context.next = 9;
          break;
        case 8:
          debug.warn('You seem to have no bookmark file, nor even a profile file!');
        case 9:
          return _context.abrupt("return", userContext);
        case 10:
        case "end":
          return _context.stop();
      }
    }, _callee, null, [[3, 5]]);
  }));
  return _findBookmarkDocument.apply(this, arguments);
}
function addBookmark(_x2, _x3) {
  return _addBookmark.apply(this, arguments);
}
function _addBookmark() {
  _addBookmark = (0, _asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function _callee2(context, target) {
    var title, me, author, bookmarkDoc, bookmark, ins, msg, _t2;
    return _regenerator["default"].wrap(function (_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          /* like
          @prefix terms: <http://purl.org/dc/terms/>.
          @prefix bookm: <http://www.w3.org/2002/01/bookmark#>.
          @prefix n0: <http://xmlns.com/foaf/0.1/>.
          <> terms:references <#0.5534145389246576>.
          <#0.5534145389246576>
           a bookm:Bookmark;
           terms:created "2019-01-26T20:26:44.374Z"^^XML:dateTime;
           terms:title "Herons";
           bookm:recalls wiki:Heron;
           n0:maker c:me.
          */
          title = '';
          me = _solidLogic.authn.currentUser(); // If already logged on
          if (me) {
            _context2.next = 1;
            break;
          }
          throw new Error('Must be logged on to add Bookmark');
        case 1:
          author = _solidLogic.store.any(target, ns.foaf('maker'));
          title = label(author) + ': ' + _solidLogic.store.anyValue(target, ns.sioc('content')).slice(0, 80); // @@ add chat title too?
          bookmarkDoc = context.bookmarkDocument;
          bookmark = UI.widgets.newThing(bookmarkDoc, title);
          ins = [$rdf.st(bookmarkDoc, UI.ns.dct('references'), bookmark, bookmarkDoc), $rdf.st(bookmark, UI.ns.rdf('type'), BOOK('Bookmark'), bookmarkDoc), $rdf.st(bookmark, UI.ns.dct('created'), new Date(), bookmarkDoc), $rdf.st(bookmark, BOOK('recalls'), target, bookmarkDoc), $rdf.st(bookmark, UI.ns.foaf('maker'), me, bookmarkDoc), $rdf.st(bookmark, UI.ns.dct('title'), title, bookmarkDoc)];
          _context2.prev = 2;
          _context2.next = 3;
          return updatePromise([], ins);
        case 3:
          _context2.next = 5;
          break;
        case 4:
          _context2.prev = 4;
          _t2 = _context2["catch"](2);
          msg = 'Making bookmark: ' + _t2;
          debug.warn(msg);
          return _context2.abrupt("return", null);
        case 5:
          return _context2.abrupt("return", bookmark);
        case 6:
        case "end":
          return _context2.stop();
      }
    }, _callee2, null, [[2, 4]]);
  }));
  return _addBookmark.apply(this, arguments);
}
function toggleBookmark(_x4, _x5, _x6) {
  return _toggleBookmark.apply(this, arguments);
}
function _toggleBookmark() {
  _toggleBookmark = (0, _asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function _callee3(userContext, target, bookmarkButton) {
    var bookmarks, i, bookmark, _t3;
    return _regenerator["default"].wrap(function (_context3) {
      while (1) switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 1;
          return _solidLogic.store.fetcher.load(userContext.bookmarkDocument);
        case 1:
          bookmarks = _solidLogic.store.each(null, BOOK('recalls'), target, userContext.bookmarkDocument);
          if (!bookmarks.length) {
            _context3.next = 9;
            break;
          }
          if (confirm('Delete bookmark on this?' + bookmarks.length)) {
            _context3.next = 2;
            break;
          }
          return _context3.abrupt("return");
        case 2:
          i = 0;
        case 3:
          if (!(i < bookmarks.length)) {
            _context3.next = 8;
            break;
          }
          _context3.prev = 4;
          _context3.next = 5;
          return updatePromise(_solidLogic.store.connectedStatements(bookmarks[i]), []);
        case 5:
          bookmarkButton.style.backgroundColor = 'white';
          debug.log('Bookmark deleted: ' + bookmarks[i]);
          _context3.next = 7;
          break;
        case 6:
          _context3.prev = 6;
          _t3 = _context3["catch"](4);
          debug.error('Cant delete bookmark:' + _t3);
          debug.warn('Cannot delete bookmark:' + _t3);
        case 7:
          i++;
          _context3.next = 3;
          break;
        case 8:
          _context3.next = 11;
          break;
        case 9:
          _context3.next = 10;
          return addBookmark(userContext, target);
        case 10:
          bookmark = _context3.sent;
          bookmarkButton.style.backgroundColor = 'yellow';
          debug.log('Bookmark added: ' + bookmark);
        case 11:
        case "end":
          return _context3.stop();
      }
    }, _callee3, null, [[4, 6]]);
  }));
  return _toggleBookmark.apply(this, arguments);
}
function renderBookmarksButton(_x7, _x8) {
  return _renderBookmarksButton.apply(this, arguments);
}
function _renderBookmarksButton() {
  _renderBookmarksButton = (0, _asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function _callee5(userContext, target) {
    var setBookmarkButtonColor, _setBookmarkButtonColor, bookmarkButton;
    return _regenerator["default"].wrap(function (_context5) {
      while (1) switch (_context5.prev = _context5.next) {
        case 0:
          _setBookmarkButtonColor = function _setBookmarkButtonCol2() {
            _setBookmarkButtonColor = (0, _asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function _callee4(bookmarkButton) {
              var bookmarked;
              return _regenerator["default"].wrap(function (_context4) {
                while (1) switch (_context4.prev = _context4.next) {
                  case 0:
                    _context4.next = 1;
                    return _solidLogic.store.fetcher.load(userContext.bookmarkDocument);
                  case 1:
                    bookmarked = _solidLogic.store.any(null, BOOK('recalls'), bookmarkButton.target, userContext.bookmarkDocument);
                    bookmarkButton.style = UI.style.buttonStyle;
                    if (bookmarked) bookmarkButton.style.backgroundColor = 'yellow';
                  case 2:
                  case "end":
                    return _context4.stop();
                }
              }, _callee4);
            }));
            return _setBookmarkButtonColor.apply(this, arguments);
          };
          setBookmarkButtonColor = function _setBookmarkButtonCol(_x9) {
            return _setBookmarkButtonColor.apply(this, arguments);
          };
          if (!userContext.bookmarkDocument) {
            _context5.next = 2;
            break;
          }
          bookmarkButton = UI.widgets.button(dom, UI.icons.iconBase + BOOKMARK_ICON, label(BOOK('Bookmark')), function () {
            toggleBookmark(userContext, target, bookmarkButton);
          });
          bookmarkButton.target = target;
          _context5.next = 1;
          return setBookmarkButtonColor(bookmarkButton);
        case 1:
          return _context5.abrupt("return", bookmarkButton);
        case 2:
        case "end":
          return _context5.stop();
      }
    }, _callee5);
  }));
  return _renderBookmarksButton.apply(this, arguments);
}
//# sourceMappingURL=bookmarks.js.map