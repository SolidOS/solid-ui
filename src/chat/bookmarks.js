/**
 * Functions related to chat and bookmarks
 * @packageDocumentation
 */

/* global alert confirm */

import * as debug from '../debug'

const UI = {
  authn: require('../authn/authn'),
  icons: require('../iconBase'),
  ns: require('../ns'),
  media: require('../media-capture'),
  pad: require('../pad'),
  rdf: require('rdflib'),
  store: require('../store'),
  style: require('../style'),
  utils: require('../utils'),
  widgets: require('../widgets')
}
const $rdf = UI.rdf

const BOOK = $rdf.Namespace('http://www.w3.org/2002/01/bookmark#')
const BOOKMARK_ICON = 'noun_45961.svg'

const kb = UI.store
const ns = UI.ns
const label = UI.utils.label
const dom = UI.dom || window.document

/** Create a resource if it really does not exist
 *  Be absolutely sure something does not exist before creating a new empty file
 * as otherwise existing could  be deleted.
 * @param doc {NamedNode} - The resource
 */
function createIfNotExists (doc) {
  return new Promise(function (resolve, reject) {
    kb.fetcher.load(doc).then(
      response => {
        debug.log('createIfNotExists doc exists, all good ' + doc)
        // kb.fetcher.webOperation('HEAD', doc.uri).then(response => {
        resolve(response)
      },
      err => {
        if (err.response.status === 404) {
          debug.log(
            'createIfNotExists doc does NOT exist, will create... ' + doc
          )

          kb.fetcher
            .webOperation('PUT', doc.uri, {
              data: '',
              contentType: 'text/turtle'
            })
            .then(
              response => {
                // fetcher.requested[doc.uri] = 'done' // do not need to read ??  but no headers
                delete kb.fetcher.requested[doc.uri] // delete cached 404 error
                debug.log('createIfNotExists doc created ok ' + doc)
                resolve(response)
              },
              err => {
                debug.log('createIfNotExists doc FAILED: ' + doc + ': ' + err)
                reject(err)
              }
            )
        } else {
          debug.log(
            'createIfNotExists doc load error NOT 404:  ' + doc + ': ' + err
          )
          reject(err)
        }
      }
    )
  })
}

// @@@@ use the one in rdflib.js when it is avaiable and delete this
function updatePromise (del, ins) {
  return new Promise(function (resolve, reject) {
    kb.updater.update(del, ins, function (uri, ok, errorBody) {
      if (!ok) {
        reject(new Error(errorBody))
      } else {
        resolve()
      }
    }) // callback
  }) // promise
}

// export findBookmarkDocument,

/*         Bookmarking
 */
/** Find a user's bookmarks
 */
export async function findBookmarkDocument (userContext) {
  const theClass = BOOK('Bookmark')
  const fileTail = 'bookmarks.ttl'
  const isPublic = true

  await UI.authn.findAppInstances(userContext, theClass, isPublic) // public -- only look for public links
  if (userContext.instances && userContext.instances.length > 0) {
    userContext.bookmarkDocument = userContext.instances[0]
    if (userContext.instances.length > 1) {
      alert('More than one bookmark file! ' + userContext.instances)
    }
  } else {
    if (userContext.publicProfile) {
      // publicProfile or preferencesFile
      var newBookmarkFile = $rdf.sym(
        userContext.publicProfile.dir().uri + fileTail
      )
      try {
        debug.log('Creating new bookmark file ' + newBookmarkFile)
        await createIfNotExists(newBookmarkFile)
      } catch (e) {
        alert.error("Can't make fresh bookmark file:" + e)
        return userContext
      }
      await UI.authn.registerInTypeIndex(
        userContext,
        newBookmarkFile,
        theClass,
        true
      ) // public
      userContext.bookmarkDocument = newBookmarkFile
    } else {
      alert('You seem to have no bookmark file and not even a profile file.')
    }
  }
  return userContext
}

/** Add a bookmark
 */

async function addBookmark (context, target) {
  /* like
 @prefix terms: <http://purl.org/dc/terms/>.
 @prefix bookm: <http://www.w3.org/2002/01/bookmark#>.
 @prefix n0: <http://xmlns.com/foaf/0.1/>.
 <> terms:references <#0.5534145389246576>.
 <#0.5534145389246576>
   a bookm:Bookmark;
   terms:created "2019-01-26T20:26:44.374Z"^^XML:dateTime;
   terms:title "Herons";
   bookm:recalls wiki:Heron;
   n0:maker c:me.
  */
  var title = ''
  var me = UI.authn.currentUser() // If already logged on
  if (!me) throw new Error('Must be logged on to add Bookmark')

  var author = kb.any(target, ns.foaf('maker'))
  title =
    label(author) + ': ' + kb.anyValue(target, ns.sioc('content')).slice(0, 80) // @@ add chat title too?
  const bookmarkDoc = context.bookmarkDocument
  const bookmark = UI.widgets.newThing(bookmarkDoc, title)
  const ins = [
    $rdf.st(bookmarkDoc, UI.ns.dct('references'), bookmark, bookmarkDoc),
    $rdf.st(bookmark, UI.ns.rdf('type'), BOOK('Bookmark'), bookmarkDoc),
    $rdf.st(bookmark, UI.ns.dct('created'), new Date(), bookmarkDoc),
    $rdf.st(bookmark, BOOK('recalls'), target, bookmarkDoc),
    $rdf.st(bookmark, UI.ns.foaf('maker'), me, bookmarkDoc),
    $rdf.st(bookmark, UI.ns.dct('title'), title, bookmarkDoc)
  ]
  try {
    await updatePromise([], ins) // 20190118A
  } catch (e) {
    const msg = 'Making bookmark: ' + e
    alert.error(msg)
    return null
  }
  return bookmark
}

export async function toggleBookmark (userContext, target, bookmarkButton) {
  await kb.fetcher.load(userContext.bookmarkDocument)
  const bookmarks = kb.each(
    null,
    BOOK('recalls'),
    target,
    userContext.bookmarkDocument
  )
  if (bookmarks.length) {
    // delete
    if (!confirm('Delete bookmark on this?' + bookmarks.length)) return
    for (let i = 0; i < bookmarks.length; i++) {
      try {
        await updatePromise(kb.connectedStatements(bookmarks[i]), [])
        bookmarkButton.style.backgroundColor = 'white'
        debug.log('Bookmark deleted: ' + bookmarks[i])
      } catch (e) {
        debug.error('Cant delete bookmark:' + e)
        alert('Cant delete bookmark:' + e)
      }
    }
  } else {
    const bookmark = await addBookmark(userContext, target)
    bookmarkButton.style.backgroundColor = 'yellow'
    debug.log('Bookmark added: ' + bookmark)
  }
}

export async function renderBookmarksButton (userContext, target) {
  async function setBookmarkButtonColor (bookmarkButton) {
    await kb.fetcher.load(userContext.bookmarkDocument)
    const bookmarked = kb.any(
      null,
      BOOK('recalls'),
      bookmarkButton.target,
      userContext.bookmarkDocument
    )
    bookmarkButton.style = UI.style.buttonStyle
    if (bookmarked) bookmarkButton.style.backgroundColor = 'yellow'
  }

  var bookmarkButton
  if (userContext.bookmarkDocument) {
    bookmarkButton = UI.widgets.button(
      dom,
      UI.icons.iconBase + BOOKMARK_ICON,
      label(BOOK('Bookmark')),
      () => {
        toggleBookmark(userContext, target, bookmarkButton)
      }
    )
    bookmarkButton.target = target
    await setBookmarkButtonColor(bookmarkButton)
    return bookmarkButton
  }
}
