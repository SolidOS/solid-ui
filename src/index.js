/*
The MIT License (MIT)

Copyright (c) 2015-2016 Solid

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


If you would like to know more about the solid Solid project, please see
https://github.com/solid/solid
*/
'use strict'
/**
 * Provides a Solid client helper object (which exposes various static modules).
 * @module solidUi.js
 * @main solidUi.js
 */

/**
 * @class SolidUi
 * @static
 */

var rdf = require('rdflib') // pull in first avoid cross-refs
var ns = require('./ns')

var UI = {
  acl: require('./acl'),
  aclControl: require('./acl-control'),
  icons: require('./iconBase'),
  log: require('./log'),
  matrix: require('./matrix'),
  messageArea: require('./messageArea'),
  ns: ns,
  pad: require('./pad'),
  rdf:  rdf,
  signin: require('./signin'),
  solid: require('solid-client'),
  store: require('./store'),
  table: require('./table'),
  tabs: require('./tabs'),
  utils: require('./utils'),
  widgets: require('./widgets')
}

module.exports = UI
