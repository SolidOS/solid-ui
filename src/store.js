// This module of solid-ui has a main quadstore for the app to use
//

import * as debug from './debug'

var rdf = require('rdflib')
var store = (module.exports = rdf.graph()) // Make a Quad store
rdf.fetcher(store) // Attach a web I/O module, store.fetcher
store.updater = new rdf.UpdateManager(store) // Add real-time live updates store.updater

debug.log('Unique quadstore initialized.')

// ends
