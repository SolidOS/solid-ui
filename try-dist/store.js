"use strict";

// This module of solid-ui has a main quadstore for the app to use
//
var rdf = require('rdflib');

var store = module.exports = rdf.graph(); // Make a Quad store

rdf.fetcher(store); // Attach a web I/O module, store.fetcher

store.updater = new rdf.UpdateManager(store); // Add real-time live updates store.updater

console.log('Unique quadstore initialized.'); // ends
//# sourceMappingURL=store.js.map