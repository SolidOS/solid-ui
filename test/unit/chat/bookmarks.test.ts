import { silenceDebugMessages } from '../../helpers/setup'
import { findBookmarkDocument, toggleBookmark, renderBookmarksButton } from '../../../src/chat/bookmarks'
import { NamedNode, Namespace } from 'rdflib'
import store from '../../../src/store'
import { clearStore } from '../helpers/clearStore'
import { ns } from '../../../src/'

const BOOK = Namespace('http://www.w3.org/2002/01/bookmark#')

silenceDebugMessages()
jest.mock('solid-auth-client')

describe('findBookmarkDocument', () => {
  afterEach(clearStore)

  it('exists', () => {
    expect(findBookmarkDocument).toBeInstanceOf(Function)
  })

  // findBookmarkDocument â€º runs
  //   TypeError: thisIndex.map(...).flat is not a function
  it('runs', async () => {
    const context = {
      index: {}
    }
    window.alert = () => {}

    const result = await findBookmarkDocument(context)
    expect(result).toEqual({
      containers: [],
      index: {
        public: []
      },
      instances: []
    })
  })
  it('complains if you have multiple bookmark docs', async () => {
    const privReg = new NamedNode('http://example.com/privType.ttl#reg1')
    const pubReg1 = new NamedNode('http://example.com/pubType.ttl#reg1')
    const pubReg2 = new NamedNode('http://example.com/pubType.ttl#reg2')
    const bookmarksDoc1 = new NamedNode('http://example.com/bookmarks1.ttl')
    const bookmarksDoc2 = new NamedNode('http://example.com/bookmarks2.ttl')
    const bookmarksDoc3 = new NamedNode('http://example.com/bookmarks3.ttl')
    const context = {
      me: new NamedNode('http://example.com/profile/card#me'),
      publicProfile: new NamedNode('http://example.com/profile/card'),
      preferencesFile: new NamedNode('http://example.com/prefs.ttl'),
      privateTypeIndex: new NamedNode('http://example.com/privType.ttl'),
      publicTypeIndex: new NamedNode('http://example.com/pubType.ttl')
    }

    window.alert = jest.fn()
    store.add(context.me, ns.space('preferencesFile'), context.preferencesFile, context.publicProfile)
    // announce private index in settings:
    store.add(context.me, ns.solid('privateTypeIndex'), context.privateTypeIndex, context.preferencesFile)
    // announce public index in profile:
    store.add(context.me, ns.solid('publicTypeIndex'), context.publicTypeIndex, context.publicProfile)

    store.add(privReg, ns.rdf('type'), ns.solid('TypeRegistration'), context.privateTypeIndex)
    store.add(privReg, ns.solid('forClass'), BOOK('Bookmark'), context.privateTypeIndex)
    store.add(privReg, ns.solid('instance'), bookmarksDoc1, context.privateTypeIndex)

    store.add(pubReg1, ns.rdf('type'), ns.solid('TypeRegistration'), context.publicTypeIndex)
    store.add(pubReg1, ns.solid('forClass'), BOOK('Bookmark'), context.publicTypeIndex)
    store.add(pubReg1, ns.solid('instance'), bookmarksDoc2, context.publicTypeIndex)

    store.add(pubReg2, ns.rdf('type'), ns.solid('TypeRegistration'), context.publicTypeIndex)
    store.add(pubReg2, ns.solid('forClass'), BOOK('Bookmark'), context.publicTypeIndex)
    store.add(pubReg2, ns.solid('instance'), bookmarksDoc3, context.publicTypeIndex)

    const result = await findBookmarkDocument(context)
    expect((window.alert as any).mock.calls.length).toEqual(1)
    expect((window.alert as any).mock.calls[0][0]).toMatchSnapshot()
  })
  it('complains if you have multiple bookmark docs', async () => {
    const privReg = new NamedNode('http://example.com/privType.ttl#reg1')
    const pubReg1 = new NamedNode('http://example.com/pubType.ttl#reg1')
    const pubReg2 = new NamedNode('http://example.com/pubType.ttl#reg2')
    const bookmarksDoc1 = new NamedNode('http://example.com/bookmarks1.ttl')
    const bookmarksDoc2 = new NamedNode('http://example.com/bookmarks2.ttl')
    const bookmarksDoc3 = new NamedNode('http://example.com/bookmarks3.ttl')
    const context = {
      me: new NamedNode('http://example.com/profile/card#me'),
      publicProfile: new NamedNode('http://example.com/profile/card'),
      preferencesFile: new NamedNode('http://example.com/prefs.ttl'),
      privateTypeIndex: new NamedNode('http://example.com/privType.ttl'),
      publicTypeIndex: new NamedNode('http://example.com/pubType.ttl')
    }

    window.alert = jest.fn()
    store.add(context.me, ns.space('preferencesFile'), context.preferencesFile, context.publicProfile)
    // announce private index in settings:
    store.add(context.me, ns.solid('privateTypeIndex'), context.privateTypeIndex, context.preferencesFile)
    // announce public index in profile:
    store.add(context.me, ns.solid('publicTypeIndex'), context.publicTypeIndex, context.publicProfile)

    store.add(privReg, ns.rdf('type'), ns.solid('TypeRegistration'), context.privateTypeIndex)
    store.add(privReg, ns.solid('forClass'), BOOK('Bookmark'), context.privateTypeIndex)
    store.add(privReg, ns.solid('instance'), bookmarksDoc1, context.privateTypeIndex)

    store.add(pubReg1, ns.rdf('type'), ns.solid('TypeRegistration'), context.publicTypeIndex)
    store.add(pubReg1, ns.solid('forClass'), BOOK('Bookmark'), context.publicTypeIndex)
    store.add(pubReg1, ns.solid('instance'), bookmarksDoc2, context.publicTypeIndex)

    store.add(pubReg2, ns.rdf('type'), ns.solid('TypeRegistration'), context.publicTypeIndex)
    store.add(pubReg2, ns.solid('forClass'), BOOK('Bookmark'), context.publicTypeIndex)
    store.add(pubReg2, ns.solid('instance'), bookmarksDoc3, context.publicTypeIndex)

    const result = await findBookmarkDocument(context)
    expect((window.alert as any).mock.calls.length).toEqual(1)
    expect((window.alert as any).mock.calls[0][0]).toMatchSnapshot()
  })
  it('does not complains if you have one bookmark docs', async () => {
    const privReg = new NamedNode('http://example.com/privType.ttl#reg1')
    const pubReg1 = new NamedNode('http://example.com/pubType.ttl#reg1')
    const pubReg2 = new NamedNode('http://example.com/pubType.ttl#reg2')
    const bookmarksDoc1 = new NamedNode('http://example.com/bookmarks1.ttl')
    const bookmarksDoc2 = new NamedNode('http://example.com/bookmarks2.ttl')
    const bookmarksDoc3 = new NamedNode('http://example.com/bookmarks3.ttl')
    const context = {
      me: new NamedNode('http://example.com/profile/card#me'),
      publicProfile: new NamedNode('http://example.com/profile/card'),
      preferencesFile: new NamedNode('http://example.com/prefs.ttl'),
      privateTypeIndex: new NamedNode('http://example.com/privType.ttl'),
      publicTypeIndex: new NamedNode('http://example.com/pubType.ttl')
    }

    window.alert = jest.fn()
    store.add(context.me, ns.space('preferencesFile'), context.preferencesFile, context.publicProfile)
    // announce private index in settings:
    store.add(context.me, ns.solid('privateTypeIndex'), context.privateTypeIndex, context.preferencesFile)
    // announce public index in profile:
    store.add(context.me, ns.solid('publicTypeIndex'), context.publicTypeIndex, context.publicProfile)

    store.add(privReg, ns.rdf('type'), ns.solid('TypeRegistration'), context.privateTypeIndex)
    store.add(privReg, ns.solid('forClass'), BOOK('Bookmark'), context.privateTypeIndex)
    store.add(privReg, ns.solid('instance'), bookmarksDoc1, context.privateTypeIndex)

    store.add(pubReg1, ns.rdf('type'), ns.solid('TypeRegistration'), context.publicTypeIndex)
    store.add(pubReg1, ns.solid('forClass'), BOOK('Bookmark'), context.publicTypeIndex)
    store.add(pubReg1, ns.solid('instance'), bookmarksDoc2, context.publicTypeIndex)

    const result = await findBookmarkDocument(context)
    expect((window.alert as any).mock.calls.length).toEqual(0)
  })
})

describe('toggleBookmark', () => {
  it('exists', () => {
    expect(toggleBookmark).toBeInstanceOf(Function)
  })
  it('runs', async () => {
    const userContext = {
      me: new NamedNode('http://example.com'),
      bookmarkDocument: new NamedNode('http://example.com')
    }
    const target = {}
    const bookmarkButton = {}
    try {
      await toggleBookmark(userContext, target, bookmarkButton)
    } catch (e) {
      expect(e.message).toEqual('Must be logged on to add Bookmark')
    }
  })
})

describe('renderBookmarksButton', () => {
  it('exists', () => {
    expect(renderBookmarksButton).toBeInstanceOf(Function)
  })
  it('runs', async () => {
    const userContext = {
      me: new NamedNode('http://example.com'),
      bookmarkDocument: new NamedNode('http://example.com')
    }
    const target = {}
    const result = await renderBookmarksButton(userContext, target)
    expect(typeof result).toEqual('object')
  })
})
