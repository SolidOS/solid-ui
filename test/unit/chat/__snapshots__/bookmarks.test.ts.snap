// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findBookmarkDocument complains if you have multiple bookmark docs 1`] = `"More than one bookmark file! <http://example.com/bookmarks2.ttl>,<http://example.com/bookmarks3.ttl>"`;

exports[`findBookmarkDocument complains if you have multiple bookmark docs 2`] = `"More than one bookmark file! <http://example.com/bookmarks2.ttl>,<http://example.com/bookmarks3.ttl>"`;
